File: apps/web/lib/thought_sync.ts
/** 想法云同步 outbox（apply 在 sync.ts） */

import { enqueue, type Envelope } from './sync';
import { userLsGet, userLsSet } from './user_storage';

const VER_MAP_KEY = 'thought_sync_versions_v1';

function readVers(): Record<string, number> {
  try {
    return JSON.parse(userLsGet(VER_MAP_KEY) || '{}');
  } catch {
    return {};
  }
}

function writeVers(m: Record<string, number>) {
  userLsSet(VER_MAP_KEY, JSON.stringify(m));
}

function bumpVersion(id: string): number {
  const vers = readVers();
  const next = (vers[id] ?? 0) + 1;
  vers[id] = next;
  writeVers(vers);
  return next;
}

export function remoteVersionForThought(id: string): number {
  return readVers()[id] ?? 0;
}

export function recordRemoteThought(id: string, version: number) {
  const vers = readVers();
  vers[id] = version;
  writeVers(vers);
}

export function clearThoughtSyncMeta(id: string) {
  const vers = readVers();
  delete vers[id];
  writeVers(vers);
}

export function enqueueThought(opts: {
  id: string;
  ref: string;
  body: string;
  visibility: string;
  createdAtMs: number;
  isDelete?: boolean;
}) {
  const version = bumpVersion(opts.id);
  const env: Envelope = {
    entity: 'thought',
    op: opts.isDelete ? 'delete' : 'update',
    id: opts.id,
    version,
    client_ts: new Date().toISOString(),
    ...(opts.isDelete
      ? {}
      : {
          data: {
            ref: opts.ref,
            body: opts.body,
            visibility: opts.visibility,
            created_at_ms: opts.createdAtMs,
          },
        }),
  };
  enqueue(env);
  if (opts.isDelete) clearThoughtSyncMeta(opts.id);
}

/** 确保本地想法 id 为 UUID，便于服务端主键 */
export function ensureThoughtSyncId(existing?: string): string {
  if (
    existing &&
    /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(
      existing,
    )
  ) {
    return existing;
  }
  if (typeof crypto !== 'undefined' && crypto.randomUUID) return crypto.randomUUID();
  return `00000000-0000-4000-8000-${String(Date.now()).padStart(12, '0').slice(-12)}`;
}

File: apps/mobile/lib/features/bible/thought_sync.dart
/// 想法云同步 outbox（对齐 Web `thought_sync.ts`）。
library;

import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/api_client.dart' show prefsProvider;
import '../../core/user_storage.dart';
import '../notes/notes_repository.dart' show syncEngineProvider;

const _verMapKey = 'thought_sync_versions_v1';

Map<String, int> _readVers(SharedPreferences prefs) {
  try {
    final raw = userPrefGetString(prefs, _verMapKey);
    if (raw == null || raw.isEmpty) return {};
    return (jsonDecode(raw) as Map).map(
      (k, v) => MapEntry(k as String, (v as num).toInt()),
    );
  } catch (_) {
    return {};
  }
}

Future<void> _writeVers(SharedPreferences prefs, Map<String, int> m) async {
  await userPrefSetString(prefs, _verMapKey, jsonEncode(m));
}

int bumpThoughtVersion(SharedPreferences prefs, String id) {
  final vers = _readVers(prefs);
  final next = (vers[id] ?? 0) + 1;
  vers[id] = next;
  userPrefSetString(prefs, _verMapKey, jsonEncode(vers));
  return next;
}

int remoteVersionForThought(SharedPreferences prefs, String id) =>
    _readVers(prefs)[id] ?? 0;

Future<void> recordRemoteThought(
  SharedPreferences prefs,
  String id,
  int version,
) async {
  final vers = _readVers(prefs);
  vers[id] = version;
  await _writeVers(prefs, vers);
}

Future<void> clearThoughtSyncMeta(SharedPreferences prefs, String id) async {
  final vers = _readVers(prefs);
  vers.remove(id);
  await _writeVers(prefs, vers);
}

final _uuid = Uuid();

/// 确保本地想法 id 为 UUID，便于服务端主键。
String ensureThoughtSyncId(String? existing) {
  if (existing != null &&
      RegExp(
        r'^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$',
        caseSensitive: false,
      ).hasMatch(existing)) {
    return existing;
  }
  return _uuid.v4();
}

Future<void> enqueueThoughtSync(
  Ref ref, {
  required String id,
  required String refStr,
  required String body,
  required String visibility,
  required int createdAtMs,
  bool isDelete = false,
}) async {
  final prefs = ref.read(prefsProvider);
  final version = bumpThoughtVersion(prefs, id);
  await ref.read(syncEngineProvider).enqueueThought(
        id: id,
        version: version,
        refStr: refStr,
        body: body,
        visibility: visibility,
        createdAtMs: createdAtMs,
        isDelete: isDelete,
      );
  if (isDelete) await clearThoughtSyncMeta(prefs, id);
}
